strawberryfields/engine.py

```diff
@@ -340,4 +346,127 @@ def run(self, program, *, shots=1, compile_options={}, modes=None, state_options
+    the results available via :class:`.Result`.
+
+    Args:
+        backend (str, BaseBackend): name of the backend, or a pre-constructed backend instance
+    """
+
+    API_DEFAULT_REFRESH_SECONDS = 0
+    HARDWARE_BACKENDS = ('chip0', )
+
+    def __init__(self):
+        # Only chip0 backend supported initially.
+        backend = "chip0"
+        super().__init__(backend)
+
+        self.client = APIClient(hostname="localhost")
+        self.jobs = []
+
+    def __str__(self):
+        return self.__class__.__name__ + '({})'.format(self.backend_name)
+
+    def reset(self, backend_options=None):
+        """
+        Reset must be called in order to submit a new job. This clears the job queue as well as
+        any ran Programs.
+        """
+        if backend_options is None:
+            backend_options = {}
+
+        super().reset(backend_options)
+        self.jobs.clear()
+
+    def _init_backend(self, *args):
+        """
+        TODO: This does not do anything rightn now.
+        """
+        # Do nothing for now...
+        pass
+
+    def generate_job_content(self, name, shots, blackbird_code):
+        """
+        Generates a string representing the Blackbird code that will be sent to the server.
+        Assumes the current backend as the target.
+
+        Args:
+            name (str): The name of the job to be created (e.g. StateTeleportation).
+            shots (int): The number of shots.
+            blackbird_code: The blackbird code of the job.
+
+        Returns:
+            str: A string containing the job content to be sent to the server.
+        """
+        target = self.backend_name
+        template = """
+            name {name}
+            version 1.0
+            target {target} (shots={shots})
+
+            {blackbird_code}
+        """.format(
+                name=name,
+                target=target,
+                shots=str(shots),
+                blackbird_code=blackbird_code)
+
+        return "\n".join([l.strip() for l in template.strip().split("\n")])
+
+    def _run_program(self, program, **kwargs):
+        """
+        Given a compiled program, gets the blackbird circuit code and creates (or resumes) a job
+        via the API. If the job is completed, returns the job result.
+
+        A queued job can be interrupted by a KeyboardInterrupt event, at which point if the job ID
+        was retrieved from the server, the job will be accessible via engine.jobs.
+
+        Args:
+            program (strawberryfields.program.Program): A program instance to be executed remotely.
+
+        Returns:
+            (list): A list representing the result samples
+
+        Raises:
+            Exception: In case a job could not be submitted or completed.
+            TypeError: In case a job is already queued and a user is trying to submit a new job.
+        """
+        blackbird_code = program.get_blackbird_syntax()
```

Comment:
Note: rather than modifying `program.py` and `ops.py` to print out the Blackbird command, it would be better to use the `strawberryfields.io` module:

```python
from strawberryfields.io import to_blackbird
blackbird_code = to_blackbird(program).serialize()
```

Mainly because it avoids code duplication, is already well tested, and takes into account lots of edge cases (such as array parameters, measurements, register transforms, etc.)

Aside: should we make a `Program.to_blackbird` method that calls the above function?

Reply:
@josh146 Thanks! Did not look hard enough for this `to_blackbird` method, I will switch this.

Reply:
@josh146 updated.